src/Network/NWEthernetChannel.cs

```diff
@@ -0,0 +1,208 @@
+		static void TrampolineReceiveHandler (IntPtr block, OS_dispatch_data content, ushort vlanTag, byte[] localAddress, byte[] remoteAddress)
+		{
+			var del = BlockLiteral.GetTarget<NWEthernetChannelReceiveDelegate> (block);
+			if (del != null) {
+				DispatchData dispatchData = null, dataCopy = null;
+				IntPtr bufferAddress = IntPtr.Zero;
+				nuint bufferSize = 0;
+
+				if (content != IntPtr.Zero) {
+					dispatchData = new DispatchData (content, owns: false);
+					dataCopy = dispatchData.CreateMap (out bufferAddress, out bufferSize);
+				}
+
+				unsafe {
+					var spanData = new ReadOnlySpan<byte> ((void*)bufferAddress, (int)bufferSize);
+					del (spanData, vlanTag, Encoding.UTF8.GetString (localAddress), Encoding.UTF8.GetString (remoteAddress));
+				}
+
+				if (dispatchData != null) {
+					dataCopy.Dispose ();
+					dispatchData.Dispose ();
+				}
+			}
+		}
+
+		[BindingImpl (BindingImplOptions.Optimizable)]
+		public void SetReceiveHandler (NWEthernetChannelReceiveDelegate handler)
+		{
+			unsafe
+			{
+				if (handler == null) {
+					nw_ethernet_channel_set_receive_handler (GetCheckedHandle (), null);
+					return;
+				}
+
+				BlockLiteral block_handler = new BlockLiteral ();
+				BlockLiteral *block_ptr_handler = &block_handler;
+				block_handler.SetupBlockUnsafe (static_ReceiveHandler, handler);
+				try {
+					nw_ethernet_channel_set_receive_handler (GetCheckedHandle (), (void*) block_ptr_handler);
+				} finally {
+					block_handler.CleanupBlock ();
+				}
+			}
+		}
+
+		[DllImport (Constants.NetworkLibrary)]
+		unsafe static extern void nw_ethernet_channel_set_state_changed_handler (OS_nw_ethernet_channel ethernet_channel, /* [NullAllowed] */ void *handler);
+
+		delegate void nw_ethernet_channel_state_changed_handler_t (IntPtr block, NWEthernetChannelState state, IntPtr error);
+		static nw_ethernet_channel_state_changed_handler_t static_StateChangesHandler = TrampolineStateChangesHandler;
+
+		[MonoPInvokeCallback (typeof (nw_ethernet_channel_state_changed_handler_t))]
+		static void TrampolineStateChangesHandler (IntPtr block, NWEthernetChannelState state, IntPtr error)
+		{
+			var del = BlockLiteral.GetTarget<Action<NWEthernetChannelState, NWError>> (block);
+			if (del != null) {
+				var nwError = (error == IntPtr.Zero)? null : new NWError (error, owns: false);
+				del (state, nwError);
+			}
+		}
+
+		[BindingImpl (BindingImplOptions.Optimizable)]
+		public void SetStateChangesHandler (Action<NWBrowserState, NWError> handler)
+		{
+			unsafe {
+				if (handler == null) {
+					nw_ethernet_channel_set_state_changed_handler (GetCheckedHandle (), null);
+					return;
+				}
+				BlockLiteral block_handler = new BlockLiteral ();
+				BlockLiteral *block_ptr_handler = &block_handler;
+				block_handler.SetupBlockUnsafe (static_StateChangesHandler, handler);
+				try {
+					nw_ethernet_channel_set_state_changed_handler (GetCheckedHandle (), (void*) block_ptr_handler);
```

Comment:
I believe this would work if the P/Invoke declaration is changed as suggested above:

```suggestion
					nw_ethernet_channel_set_state_changed_handler (GetCheckedHandle (), &block_handler);
```